draft-irtf-cfrg-vdaf.md

```diff
@@ -1190,6 +1190,252 @@ how the nonces are chosen, but security requires that the nonces be unique. See
+
+The first state transition, from `Start` to `Continued` or `Rejected`, is
+induced by the following transition rule. No messages are sent or received
+during this transition.
+
+~~~ transition
+def ping_pong_start(Vdaf,
+                    vdaf_verify_key: bytes[Vdaf.VERIFY_KEY_SIZE],
+                    is_leader: bool,
+                    agg_param: Vdaf.AggParam,
+                    nonce: bytes[Vdaf.NONCE_SIZE],
+                    public_share: bytes,
+                    host_input_share: bytes) -> State:
+    try:
+        prep_state = Vdaf.prep_init(
+            vdaf_verify_key,
+            0 if is_leader else 1,
+            agg_param,
+            nonce,
+            public_share,
+            host_input_share,
+        )
+        return Continue(prep_state)
+    except:
+        return Rejected()
+~~~
+
+If the state is `Rejected`, then processing halts. Otherwise, if the state is
+`Continued`, then processing continues using the following transition rule until
+a terminal state is reached.
+
+Let us call the initiating party the "Leader" and the responding party the
+"Helper". The high-level idea is that the Leader and Helper will take turns
+running the computation locally until input from their peer is required:
+
+* For a 1-round VDAF (e.g., Prio3 ({{prio3}})), the Leader sends its prep share
+  to the Helper, who computes the prep message locally, computes its output
+  share, then sends the prep message to the Leader. Preparation requires just
+  one round trip between the Leader and the Helper.
+
+* For a 2-round VDAF (e.g., Poplar1 ({{poplar1}})), the Leader sends its
+  first-round prep share to the Helper, who replies with the first-round prep
+  message and its second-round prep share. In the next request, the Leader
+  computes its second-round prep share locally, computes its output share, and
+  sends the second-round prep message to the Helper. Finally, the Helper
+  computes its own output share. Preparation requires two round trips.
+
+* In general, each request includes the Leader's prep share for the previous
+  round and/or the prep message for the current round; correspondingly, each
+  response consists of the prep message for the current round and the Helper's
+  prep share for the next round.
+
+The Aggregators proceed in this ping-ponging fashion until a step of the
+computation fails, indicating the report is invalid and should be rejected, or
+preparation is completed. All told there there are `ceil((Vdaf.ROUNDS+1)/2)`
+requests sent.
+
+Each message in the ping-pong protocol is structured as follows (expressed in
+TLS syntax {{!RFC8446}}):
+
+~~~
+enum {
+  initalize(0),
+  continue(1),
+  finish(2),
+  (255)
+} MessageType;
+
+struct {
+  MessageType type;
+  select (Message.type) {
+    case initialize:
+      opaque prep_share<0..2^32-1>;
+    case continue:
+      opaque prep_msg<0..2^32-1>;
+      opaque prep_share<0..2^32-1>;
+    case finish:
+      opaque prep_msg<0..2^32-1>;
+  };
```

Comment:
I would prefer adding the `finished` message type back, as otherwise DAP will need an additional discriminator bit outside the `Message` struct to indicate whether or not a helper's response includes a `Message` or not.

Reply:
True, but I think that discriminator bit is needed in DAP in any case. Having done a few passes of this, I think this is the simplest design we have. (I'd encourage you to also review the corresponding DAP PR -- perhaps you'll have an idea to simplify things a bit more.)

Reply:
I think this is OK (if slightly confusing due to similar names being used in DAP & VDAF): `finished` exists at the DAP level, as one of the possible states of a `PrepareStep`, and works as you describe; this VDAF `finish` message would be embedded in a DAP `PrepareStep` of type `continue`.

Reply:
(that said, this _is_ somewhat confusing; maybe there's a better way to state this abstraction boundary?)

Reply:
Perhaps. I sort of like what we have though now. It's at least consistent: If there is an actual message to send, then `outbound` will be `None`; otherwise `outbound != None`. This is useful in DAP because the Leader and Helper do different things in this case.